### PR TITLE
JBEHAVE-1571 Add the ability to monitor example table transforming

### PR DESCRIPTION
**Pull request by draker94**

### CODE REVIEW

File: jbehave-core/src/main/java/org/jbehave/core/model/ExampleTableMonitor.java [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package org.jbehave.core.model;
+
+import org.jbehave.core.model.ExamplesTable.TableProperties;
+
+/**
+ * Interface to monitor example table events
+ */
+public interface ExampleTableMonitor {
```
Comment (valfirst) [inline jbehave-core/src/main/java/org/jbehave/core/model/ExampleTableMonitor.java]:
```suggestion
public interface TableTransformerMonitor {
```

**Review by valfirst — CHANGES_REQUESTED** [review]

File: jbehave-core/src/main/java/org/jbehave/core/model/ExamplesTableFactory.java [hunk 2]
```diff
@@ -101,7 +102,18 @@ public void useKeywords(Keywords keywords) {
         this.keywords = keywords;
     }
 
+    public void useExampleTableMonitor(ExampleTableMonitor exampleTableMonitor) {
```
Comment (valfirst) [inline jbehave-core/src/main/java/org/jbehave/core/model/ExamplesTableFactory.java]: `ExamplesTableFactory` should be immutable, so such configuration should be moved to `Configuration` or any similar controls

File: jbehave-core/src/main/java/org/jbehave/core/model/NullExampleTableMonitor.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
+package org.jbehave.core.model;
+
+import org.jbehave.core.model.ExamplesTable.TableProperties;
+
+/**
+ * <a href="http://en.wikipedia.org/wiki/Null_Object_pattern">Null Object Pattern</a> implementation of
+ * {@link ExampleTableMonitor}. Can be extended to override only the methods of interest.
+ */
+public class NullExampleTableMonitor implements ExampleTableMonitor {
```
Comment (valfirst) [inline jbehave-core/src/main/java/org/jbehave/core/model/NullExampleTableMonitor.java]: please add also `DelegatingTableTransformerMonitor`

**Review by valfirst — CHANGES_REQUESTED** [review]
Also please do not forget to add the documentation

File: jbehave-core/src/main/java/org/jbehave/core/model/TableTransformerMonitor.java [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+import org.jbehave.core.model.ExamplesTable.TableProperties;
+
+/**
+ * Interface to monitor table transformer events
+ */
+public interface TableTransformerMonitor {
+
+    void beforeTransformerApplying(String transformerName, TableProperties properties, String inputTable);
+
+    void afterTransformerApplying(String transformerName, TableProperties properties, String outputTable);
```
Comment (uarlouski) [inline jbehave-core/src/main/java/org/jbehave/core/model/TableTransformerMonitor.java:12]: maybe something like beforeTransformation and afterTransformation?

File: jbehave-core/src/main/java/org/jbehave/core/steps/ParameterConverters.java [hunk 5]
```diff
@@ -104,6 +106,7 @@
 public class ParameterConverters {
 
     public static final StepMonitor DEFAULT_STEP_MONITOR = new SilentStepMonitor();
+    public static final TableTransformerMonitor TABLE_TRANSFORMER_MONITOR = new NullTableTransformerMonitor();
```
Comment (uarlouski) [inline jbehave-core/src/main/java/org/jbehave/core/steps/ParameterConverters.java]: why not to inline? as I see it used once

File: jbehave-core/src/main/java/org/jbehave/core/model/TableTransformerMonitor.java (hunk 4 above)
Comment (draker94) [reply jbehave-core/src/main/java/org/jbehave/core/model/TableTransformerMonitor.java:12]: > maybe something like beforeTransformation and afterTransformation?

It seems to me that the current name reflects the subject area more clearly. "beforeTransformation" can be regarded as "after applying all transformers" or in some other way

File: archetypes/jbehave-guice-archetype/src/main/resources/archetype-resources/src/main/java/MyStories.java [hunk 6]
```diff
@@ -65,11 +67,13 @@ public Configuration configuration() {
         LoadFromClasspath resourceLoader = new LoadFromClasspath(embeddableClass);
         TableTransformers tableTransformers = new TableTransformers();
         ParameterControls parameterControls = new ParameterControls();
+        // TableTransformerMonitor default instance. Other implementation may be used
+        TableTransformerMonitor transformerMonitor = new NullTableTransformerMonitor();
```
Comment (valfirst) [inline archetypes/jbehave-guice-archetype/src/main/resources/archetype-resources/src/main/java/MyStories.java]: please avoid adding `NullTableTransformerMonitor` as it doesn't bring any value here, you can just overload constructors of `ExamplesTableFactory`

Comment (valfirst) [reply archetypes/jbehave-guice-archetype/src/main/resources/archetype-resources/src/main/java/MyStories.java]: here and below

File: jbehave-core/src/main/java/org/jbehave/core/configuration/Configuration.java [hunk 7]
```diff
@@ -235,6 +237,12 @@ public abstract class Configuration {
      */
     protected StepConditionMatcher stepConditionMatcher;
 
+    /**
+     * Monitor events of example table transformers.
+     * The default implementation is {@link org.jbehave.core.model.NullTableTransformerMonitor}
```
Comment (valfirst) [inline jbehave-core/src/main/java/org/jbehave/core/configuration/Configuration.java]:
```suggestion
```

**Review by valfirst — CHANGES_REQUESTED** [review]

**Comment by valfirst** [comment]
and https://github.com/jbehave/jbehave-core/pull/53#pullrequestreview-1155355539

File: jbehave-core/src/main/java/org/jbehave/core/configuration/MostUsefulConfiguration.java [hunk 8]
```diff
@@ -64,6 +67,7 @@ public MostUsefulConfiguration() {
         useStepdocReporter(new PrintStreamStepdocReporter());
         useParanamer(new NullParanamer());
         useViewGenerator(new FreemarkerViewGenerator());
+        useTableTransformerMonitor(new NullTableTransformerMonitor());
```
Comment (valfirst) [inline jbehave-core/src/main/java/org/jbehave/core/configuration/MostUsefulConfiguration.java]:
```suggestion
```

**Review by valfirst — CHANGES_REQUESTED** [review]

File: jbehave-core/src/main/java/org/jbehave/core/configuration/MostUsefulConfiguration.java (hunk 8 above)
Comment (draker94) [reply jbehave-core/src/main/java/org/jbehave/core/configuration/MostUsefulConfiguration.java]: > 

According to the current documentation, the such elements should be used in "MostUsefulConfiguration", like "FreemarkerViewGenerator", for example

Or this statement need to be changed in the docs?
https://github.com/jbehave/jbehave-core/blob/91109074e1f8546a5e142b845c8214dbfa4b9e55/distribution/src/site/content/configuration.html#L318-L321

Comment (valfirst) [reply jbehave-core/src/main/java/org/jbehave/core/configuration/MostUsefulConfiguration.java]: I would say that's a bit outdated statement

File: distribution/src/site/content/configuration.html [hunk 9]
```diff
@@ -315,10 +315,6 @@ <h3>Scanning for steps</h3>
 
 <h3>Configurable Elements</h3>
 
-<p>All configurable elements of <a
-    href="javadoc/core/org/jbehave/core/Configuration.html">Configuration</a>
-come with a default behaviour, configured in the <a
-    href="javadoc/core/org/jbehave/core/MostUsefulConfiguration.html">MostUsefulConfiguration</a>:</p>
```
Comment (valfirst) [inline distribution/src/site/content/configuration.html]: let's do not mix up changes, despite the fact this statement should be updated, the changes are not about it

**Review by valfirst — APPROVED** [review]

**Review by uarlouski — APPROVED** [review]

**Review by ikalinin1 — APPROVED** [review]

**Pull request merged**